File: download_pmc.py
import requests
import xml.etree.ElementTree as ET
import subprocess
import urllib.parse

if __name__ == "__main__": 

    inp = input("Enter search term: ")
    search_term = urllib.parse.quote(inp)

    print(f"Searching for: {inp}")

    # Step 1: Fetch XML data from the URL
    url = f"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pmc&term={search_term}&field=title&usehistory=y"
    response = requests.get(url)

    if response.status_code != 200:
        print("Error fetching data from PMC")
        exit()

    # Step 2: Parse XML and extract IDs
    root = ET.fromstring(response.content)
    ids = [id_elem.text for id_elem in root.findall(".//Id")]

    if not ids:
        print("No IDs found in response.")
        exit()

    print(f"Retrieved {len(ids)} IDs.")
    print("IDs:", ids) 

    # Step 3: Execute AWS CLI command for each ID
    s3_bucket = "pmc-oa-opendata/oa_noncomm/txt/all" 
    s3_prefix = "PMC"

    for pmc_id in ids:
        s3_key = f"{s3_prefix}{pmc_id}.txt"
        aws_command = ["aws", "s3", "cp", f"s3://{s3_bucket}/{s3_key}", ".", "--no-sign-request", "--quiet"]

        try:
            subprocess.run(aws_command, check=True)
            print(f"Downloaded {s3_key} from S3 successfully.")
        except subprocess.CalledProcessError as e:
            print(f"Failed to download {s3_key}: {e}")

    print("Script execution completed.")

File: download_pmc.js
const axios = require("axios");
const { parseStringPromise } = require("xml2js");
const { exec } = require("node:child_process");
const pLimit = require("p-limit").default;

const SEARCH_URL =
  "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pmc&term=cobb+syndrome&field=title&usehistory=y";
const S3_BUCKET = "pmc-oa-opendata/oa_noncomm/txt/all"; // Replace with actual bucket name
const S3_PREFIX = "PMC"; // Adjust prefix if necessary

// Limit concurrent AWS CLI processes to 2
const limit = pLimit(2);

// Function to fetch XML from the URL
async function fetchArticleIDs() {
  try {
    const response = await axios.get(SEARCH_URL);
    const data = await parseStringPromise(response.data);
    const ids = data.eSearchResult.IdList[0].Id.map((id) => id);
    console.log(data.eSearchResult.IdList);
    console.log(`Fetched ${ids.length} IDs: ${ids}`);
    return ids;
  } catch (error) {
    console.error("Error fetching IDs:", error.message);
    return [];
  }
}

// Function to download a file from S3 using AWS CLI
function downloadFromS3(pmcId) {
  return new Promise((resolve, reject) => {
    const s3Key = `${S3_PREFIX}${pmcId}.txt`;
    const command = `aws s3 cp s3://${S3_BUCKET}/${s3Key} --no-sign-request .`;

    console.log(`Starting download: ${s3Key}`);
    exec(command, (error, stdout, stderr) => {
      if (error) {
        console.error(`Failed to download ${s3Key}:`, stderr);
        resolve(error);
      } else {
        console.log(`Downloaded ${s3Key} successfully.`);
        resolve(stdout);
      }
    });
  });
}

// Main function
async function main() {
  const ids = await fetchArticleIDs();
  if (ids.length === 0) return;

  // Process downloads with concurrency limit
  const downloadPromises = ids.map((id) => limit(() => downloadFromS3(id)));
  await Promise.all(downloadPromises);

  console.log("All downloads completed.");
}

main();
